### ISSUE TITLE
Convert to blocks in PHP

### ISSUE THREAD

**Issue by fisicx** [WordPress/gutenberg#13163]
A client creates posts from a CSV using a plugin that calls the wp_insert_post function.

The posts are created but all the content is in a classic editor block.

Can we have a function that automatically converts classic content into blocks. Or if there is already a way to do this programmatically where can I find the documentation?

**Comment by pbrocks** [comment]
@fisicx Thanks for the question and the opportunity to reflect on this. I believe the key point here is that a Post is still a Post and a Block is a smaller entity that is a constituent of a Post. The automatic conversion that you are referring to represents the designation of a paragraph with the Classic Block. Make sense? That is, if a post that your client is importing has 5 paragraphs, then the editor should contain 5 classic blocks after import.

Having said that, once a post is imported and living in the WordPress database, you may choose to pull that post into a block that, for example, pulls in the latest posts for display in one area of a post or page. But going back to your original question about import, if you have a desired outcome, you could wrap the csv import in the appropriate html comments to render whatever blocks you like.

**Comment by swissspidy** [comment]
> If you have a desired outcome, you could wrap the csv import in the appropriate html comments to render whatever blocks you like.

I believe OP is asking for a the same "Convert to blocks" functionality we already have in JavaScript. In other words, a function that automatically adds these HTML comments.

At the moment such a function does not exist. Mainly because block types are only registered in JavaScript and on the server you have no idea which blocks exist.

However, server-side block registration is planned as per #2751. Once that is done, we will be able to provide such a function for developers.

**Comment by fisicx** [comment]
I have looked at adding the html comments to the content but it means I've got to do this for all possible elements. And this seems redundant when there is already a function to convert classic to blocks. I can go to 'More Options > Convert to Blocks'. I just want to be able to call a function like this when the post is created. Rather than go through each of the 100+ custom post types they import from the CSV.

The people importing from a CSV are OK adding HTML tags and attributes, if I tell them they have got to learn a whole new way of doing things they won't be happy. It's this loss of simple content creation that is causing so many problems. Even more so for the thousands of plugins that use shortcodes attributes to manage content.

**Comment by swissspidy** [comment]
> And this seems redundant when there is already a function to convert classic to blocks. I can go to 'More Options > Convert to Blocks'.

That is a client-side JavaScript function, nothing that is available in PHP when importing from CSV. As I said, there is no such PHP function at the moment but this will likely be added in the future.

**Comment by fisicx** [comment]
Thanks for that. I will be patient.

In the meantime I will have to do some basic str_replace on the more common elements.

**Comment by youknowriad** [comment]
If you're adventurous, I think you can still use a node script running jsdom and use the `wp.blocks.rawHandler` function. Otherwise, you should wait for phase 2 work around server side awareness of blocks #2751

**Comment by youngmicroserf** [comment]
This might help. Came accross it while researching this issue.

https://wordpress.stackexchange.com/questions/343037/how-to-serialize-a-gutenberg-block-from-block-attributes-in-php

**Comment by pranksinatra** [comment]
Just wanted to voice my support for this issue!

In the meantime, I wrote a script that automatically opens every WP post and converts it to Gutenberg by clicking the buttons on the Edit Post page. Gross, I know, but it worked for the 900 posts on our WP site.
https://gist.github.com/pranksinatra/6573f3c8c2424556364250e04a249937

**Comment by chaselangr** [comment]
> Just wanted to voice my support for this issue!
> 
> In the meantime, I wrote a script that automatically opens every WP post and converts it to Gutenberg by clicking the buttons on the Edit Post page. Gross, I know, but it worked for the 900 posts on our WP site.
> https://gist.github.com/pranksinatra/6573f3c8c2424556364250e04a249937

is there a way to hook this action onto the publish / update action so that it just automatically does it for new posts?

**Comment by evrpress** [comment]
Until this is possible in core you should check out [convert-to-blocks](https://github.com/10up/convert-to-blocks) by @10up

**Comment by danieliser** [comment]
Why are these all closed, and after 5 years still no method to solve this, yet they still expect everybody to happily jump on the Gutenberg train like these major gaps in coverage don't exist.

**Comment by chubes4** [comment]
I created a little plugin that automatically converts html to blocks on the server side. Mirrors Gutenberg's client side rawHandler

https://github.com/chubes4/html-to-blocks-converter

**Comment by swissspidy** [comment]
@chubes4 Interesting approach converting all the JS code to PHP!

Was this a one-off thing or is there a script that could regularly do this? It might set an interesting precedent to include something like that in Gutenberg itself and make available to others.

For the HTML parsing I'd suggest using core's HTML API instead of the PHP DOM API.

**Comment by chubes4** [comment]
@swissspidy Thanks! 

It was technically a one-off thing. I was looking for the server side equivalent of "Convert to Blocks" in core and discovered it does not exist. 

However, I believe anyone can do similar things with a CLI agent and a bit of exploring. I simply downloaded Gutenberg, opened the agent inside of it, and explained what I wanted to do. After some iteration I had a working example. 

Thanks for the suggestion. I will look into giving it an update to use the core HTML API.

**Comment by swissspidy** [comment]
Makes sense, thanks. Nice to see you already made the switch. AI FTW :-)

I'm reopening this for visibility, I think there's something there that's worth exploring.

**Comment by swissspidy** [comment]
@dmsnell @youknowriad @adamziel Curious to hear your thoughts on this. (see also https://github.com/wp-cli/block-command/issues/11)

Defining block transforms in PHP could be really powerful for HTML <-> Block <-> Markdown conversions, and would be useful for imports as well (import arbitrary HTML directly into block markup).

**Comment by chubes4** [comment]
I have gone deeper with experimentation here and now have a Static Site Importer that turns raw html templates into working block themes. This proves the value of converting to blocks in PHP extends beyond post content. 

Here is a working experiment with raw HTML site generation via WordPress Studio Code:
https://github.com/Automattic/studio/pull/3309

And the Static Site Importer: https://github.com/chubes4/static-site-importer

I believe this a better approach than trying to teach AI agents how to write block markup.

**Comment by dcx15** [comment]
So to confirm, is adding this functionality to wp_insert_post() planned? Or has this issue evolved into something different from the OP's question?

**Comment by nickchomey** [comment]
Apologies if this is considered spam, but the php-only blocks work is part of what inspired me to explore allowing writing full javascript expressions directly in  interactivity api html directives - no external js file needed. 

Here's the tracking issue for it https://github.com/WordPress/gutenberg/issues/79765

Combined with the increasingly capable php only blocks, you could make fully interactive blocks directly from PHP, no JS files required (though, obviously, some JS would be needed).

**Comment by dmsnell** [comment]
@swissspidy this has been on my todo list for a long time and I just can’t seem to get the focus to attend to it properly. I’m sorry for being unavailable for it at this time.

**Comment by swissspidy** [comment]
@dmsnell no worries at all! I'll have Claude tinker with it in #82013 on the side.

**Comment by chubes4** [comment]
@swissspidy glad to see some movement here! Feel free to ping me when your experiment is ready for testing, I will be happy to take it for a spin
